### PR TITLE
Fix to testLifted

### PR DESCRIPTION
**Pull request by odersky**
If the original type does not have the right type parameters, look in the baseclasses.
Previously this was done only if the original type did not have any type parameters.

Review by @smarter

### CODE REVIEW

File: src/dotty/tools/dotc/core/TypeApplications.scala [hunk 1]
```diff
@@ -509,7 +509,7 @@ class TypeApplications(val self: Type) extends AnyVal {
    *               { type $hkArg$0 = T1; ...; type $hkArg$n = Tn }
    *
    *  satisfies predicate `p`. Try base types in the order of their occurrence in `baseClasses`.
```
Comment (smarter) [inline src/dotty/tools/dotc/core/TypeApplications.scala]: The documentation still seems incomplete, we don't always check that there is a base type where the type parameters match `tparams`, and there's a special case when `tparams.isEmpty`.

File: src/dotty/tools/dotc/typer/TypeAssigner.scala [hunk 2]
```diff
@@ -28,9 +28,18 @@ trait TypeAssigner {
     }
   }
 
-  def avoid(tp: Type, syms: => List[Symbol])(implicit ctx: Context): Type = {
+  /** An upper approximation of the given type `tp` that does not refer to any symbol in `symsToAvoid`.
```
Comment (smarter) [inline src/dotty/tools/dotc/typer/TypeAssigner.scala]: This looks fine but is unrelated to the current pull request and ideally should be split in another commit.

Comment (odersky) [inline src/dotty/tools/dotc/typer/TypeAssigner.scala]: Agreed, but I don't have time to do that. Way too many other things on my plate :-(
